util/timer.cpp

```diff
@@ -1,48 +1,63 @@
+#include <omp.h>
+#include <utility>
+#include <vector>
 #include "timer.hpp"
 
 namespace ts {
-namespace Timer {
 
 using namespace std::chrono;
 
-std::string last_func;
-auto last_time = std::chrono::high_resolution_clock::now();
+std::vector<std::map<std::string, nanoseconds>> durations_vec(100);
 
-std::map<std::string, std::chrono::nanoseconds> durations;
 
-void start(std::string func_name) {
-//    assert(last_func.empty());
-//    last_time = high_resolution_clock::now();
-//    last_func = std::move(func_name);
-    func_name = "";
-}
+void timer_add_duration(const std::string &func_name, nanoseconds dur) {
+    auto tid = omp_get_thread_num();
+    auto &durations = durations_vec[tid];
```

Comment:
please use the explicit type rather than auto. It really helps readability. Generally, only use auto when the explicit type is obvious or it's hellishly complicated.